### PR TITLE
[WIP] Preliminary PythonPackage class

### PR DESCRIPTION
**Pull request by adamjstewart**
Decided to see how hard it would be to create a `PythonPackage` class. Not too bad so far. Definitely needs a lot more testing and documentation before it's ready to merge, but I think it's a good start.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-alabaster/package.py [hunk 1]
```diff
@@ -35,9 +35,4 @@ class PyAlabaster(Package):
     version('0.7.9', 'b29646a8bbe7aa52830375b7d17b5d7a',
             url="https://pypi.python.org/packages/71/c3/70da7d8ac18a4f4c502887bd2549e05745fa403e2cd9d06a8a9910a762bc/alabaster-0.7.9.tar.gz")
 
-    extends('python')
-
     depends_on('py-setuptools', type='build')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/py-alabaster/package.py]: Why does this depend on py-setuptools?  Should that be an autodependency like with CMakePackage?  Or am I missing a distinction between setuptools and distutils?

**Comment by citibeth** [comment]
This might sound like a dumb question, but... do all packages that use the Python build system extend Python?  For example, `py-autopep8` does not; is that for real, or is it a mistake.

I believe the Spack package superclass should be based on the build system used, let's not get that confused with the contents of the package.  If somebody wants to use Python setuptools to build a C++ package, they can go right ahead...

For that reason, I think that `PythonPackage` should probably NOT auto-extend `python`.

**Comment by adamjstewart** [comment]
So in `PythonPackage`, I added:
```python
extends('python')
```
But in some packages, I needed to add something like:
```python
extends('python', ignore=r'bin/.*')
```
This is causing the documentation tests to fail with a `DirectiveError` because you can't extend a package more than once. How can I resolve this problem?

**Comment by citibeth** [comment]
Would you like to try #2281 in this PR as well; or do you think that's best left for another day?

**Comment by citibeth** [comment]
BTW... looking forward to this one!

**Comment by adamjstewart** [comment]
> do all packages that use the Python build system extend Python?

Well, all packages that need to run `python setup.py` have to _depend on_ Python. And `depends_on` == `extends`. The only difference between `extends` and `depends_on` is that you are allowed to run `spack activate py-package` and it will symlink the contents into the Python installation directory.

> For example, `py-autopep8` does not

https://github.com/LLNL/spack/blob/develop/var/spack/repos/builtin/packages/py-autopep8/package.py#L38

**Comment by adamjstewart** [comment]
> Would you like to try #2281 in this PR as well; or do you think that's best left for another day?

Perhaps. I think this PR is going to need a **lot** of testing. And like your `capitalization` PR, it affects a lot of packages, so I would like to get it merged sooner rather than later (obviously after your capitalization PR which appears to be done). But I _think_ it ought to be really easy to add a custom `url_for_version` to the `PythonPackage` base class. Or for a `PyPiPackage` base class that extends `PythonPackage`. It depends on how many Python packages are available on PyPi and how many aren't. I know `py-phonopy` and `py-meep` aren't.

By the way, `setup.py` has a lot of options:
```
$ python setup.py --name
# name of the package
$ python setup.py --version
# version of the tarball
$ python setup.py --description
# short description of the package
$ python setup.py --requires
# supposed to list dependencies, but often left blank
```
In combination with the `url_for_version`, it may be possible to fill in everything needed in a `package.py` during `spack create`.

**Comment by adamjstewart** [comment]
I'll make a similar PR for the R-build system after #2475 is merged. R seems even simpler.

File: var/spack/repos/builtin/packages/py-alabaster/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-alabaster/package.py]: I think setuptools is like a successor to distutils. The setup.py tries to import setuptools. If it is not available, we can't build it. This is the case for the majority of Python packages. I thought about adding it to PythonPackage anyway, but a lot of packages don't need it, including `py-setuptools`, for which we don't want a circular dependency.

**Comment by adamjstewart** [comment]
Ok, I have almost all of the packages that call `setup_py` converted to `PythonPackage`. I just have a couple types remaining.

The first type is like `py-rtree`, where environment variables need to be set during install:
```python
lib = os.path.join(spec['libspatialindex'].prefix, 'lib')
os.environ['SPATIALINDEX_LIBRARY'] = os.path.join(lib, 'libspatialindex.%s' % dso_suffix)                
os.environ['SPATIALINDEX_C_LIBRARY'] = os.path.join(lib, 'libspatialindex_c.%s' % dso_suffix)

setup_py('install', '--prefix=%s' % prefix)
```
Is this what `setup_environment` is for?

A second type is where there are more phases than normal, like in `py-meep`:
```python
self.setup_py('clean', '--all')
self.setup_py('build_ext', include_flags, library_flags)
self.setup_py('install', '--prefix={0}'.format(prefix))
self.setup_py('bdist', include_flags, library_flags)
```
Right now, the best idea I have is to create `<phase>` and `<phase_args>` functions for every possible phase, and then allow packages to override the `phase` list. But I'm wondering if there isn't an easier way. Could I somehow automate this process so all I need to do is override `phase` without having to create almost identical functions for every phase? Like generate the functions at runtime?

**Comment by adamjstewart** [comment]
@alalazo I'm seeing a lot of weird problems with Travis since we updated to pytest. #2711 is one example. Also, check out the last Travis run. Python 2.7 passed even though one of the tests failed!

**Comment by alalazo** [comment]
@adamjstewart @tgamblin  I think the problem you are seeing here is due to the script we run. We should change:
```shell
coverage run bin/spack test "$@"
coverage combine
```
to
```shell
coverage run bin/spack test "$@" && coverage combine
```
I'll check if I am right and and submit a PR right away in case.

File: lib/spack/spack/directives.py [hunk 2]
```diff
@@ -289,9 +289,10 @@ def extends(spec, **kwargs):
 
     """
     def _execute(pkg):
-        if pkg.extendees:
-            msg = 'Packages can extend at most one other package.'
-            raise DirectiveError(msg)
+        # if pkg.extendees:
+        #     directive = 'extends'
+        #     msg = 'Packages can extend at most one other package.'
+        #     raise DirectiveError(directive, msg)
```
Comment (adamjstewart) [inline lib/spack/spack/directives.py]: Is this safe to get rid of?

**Comment by citibeth** [comment]
> In combination with the url_for_version, it may be possible to fill in everything needed in a package.py during spack create.

Especially if the package resides on PyPI... which provides "DOAP" records giving a lot of useful info.

**Comment by adamjstewart** [comment]
Oh boy, I think every file in this PR has a conflict now. This will be fun to rebase.

**Comment by tgamblin** [comment]
On rebasing frequently: are you and @citibeth aware of [rerere](https://git-scm.com/blog/2010/03/08/rerere.html)?  It makes rebasing things with longer histories pretty painless for me.  Not sure why it's not on by default.

**Comment by adamjstewart** [comment]
@tgamblin I had never heard of that. I'll have to try it out.

While I have your mastery of git on the table, have you ever seen a message like this?
```
$ git push origin develop
Counting objects: 376, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (41/41), done.
Writing objects: 100% (376/376), 80.32 KiB | 0 bytes/s, done.
Total 376 (delta 188), reused 346 (delta 161)
remote: Resolving deltas: 100% (188/188), completed with 48 local objects.
To github.com:adamjstewart/spack.git
 ! [remote rejected] develop -> develop (pre-receive hook declined)
error: failed to push some refs to 'git@github.com:adamjstewart/spack.git'
```
It's been happening to me very frequently lately. I don't have any webhooks or protected branches on my fork. When I run `git push origin`, develop will be updated accordingly.

**Comment by citibeth** [comment]
I think I can resolve those conflicts.  Give me an hour, I'll put up a new version of this branch that will go down more easily.

**Comment by adamjstewart** [comment]
@citibeth No worries, I'm already working on rebasing. Hopefully `git rerere` will make things quicker.

**Comment by tgamblin** [comment]
@adamjstewart did you rebase and then push?  That typically requires a force-push because rebase has rewritten commits, changing the history.

If that is your problem, my advice would be to always work on branches, even within your fork, so that you only ever have to pull from upstream develop.  e.g.:

```bash
git remote add upstream <main spack repo>
git checkout -b feature
# work work work
git push origin feature
# do PR
# PR is merged
git checkout develop
git pull upstream develop
# now you're synced and feature is merged
git branch -D feature
```

**Comment by tgamblin** [comment]
It looks like you're on a branch in your fork even here, though, so I dunno why you're getting that.  Hm.

**Comment by citibeth** [comment]
OK, if you like.  My observation was that getting rid of "nolink" is
largely automated with `sed`.  I'd be happy to revert that change on the
Python files, apply your branch, and then re-apply the nolink change.  That
would probably be easier than going through a manual or semi-manual process
on your change.s

**Comment by adamjstewart** [comment]
@citibeth If that's all you have to do then that is probably easier. I tried rebasing but messed some files up, so I'll let you handle the rebase.

**Comment by adamjstewart** [comment]
@tgamblin I don't think I rebased. My develop should be in sync with your develop.

**Comment by citibeth** [comment]
Replaced by #2778, which is (mostly) a re-based version of this PR.

**Comment by tgamblin** [comment]
@adamjstewart: if you create a new fork and force push it there I could take a look...

**Comment by adamjstewart** [comment]
@tgamblin All of a sudden the problem disappeared. I wonder if my `/tmp` is full or something weird. I also filed an issue with GitHub support, so I'll see what they have to say.

**Pull request closed without merging**